acme/authority.go

```diff
@@ -263,21 +265,37 @@ func (a *Authority) ValidateChallenge(p provisioner.Interface, accID, chID strin
 	if accID != ch.getAccountID() {
 		return nil, UnauthorizedErr(errors.New("account does not own challenge"))
 	}
+	retry := ch.getRetry()
+	if retry.Active {
+		return ch.toACME(a.db, a.dir, p)
+	}
+	retry.Mux.Lock()
+	defer retry.Mux.Unlock()
+
 	client := http.Client{
 		Timeout: time.Duration(30 * time.Second),
 	}
 	dialer := &net.Dialer{
 		Timeout: 30 * time.Second,
 	}
-	ch, err = ch.validate(a.db, jwk, validateOptions{
-		httpGet:   client.Get,
-		lookupTxt: net.LookupTXT,
-		tlsDial: func(network, addr string, config *tls.Config) (*tls.Conn, error) {
-			return tls.DialWithDialer(dialer, network, addr, config)
-		},
-	})
-	if err != nil {
-		return nil, Wrap(err, "error attempting challenge validation")
+	for ch.getRetry().Active {
+		ch, err = ch.validate(a.db, jwk, validateOptions{
+			httpGet:   client.Get,
+			lookupTxt: net.LookupTXT,
+			tlsDial: func(network, addr string, config *tls.Config) (*tls.Conn, error) {
+				return tls.DialWithDialer(dialer, network, addr, config)
+			},
+		})
+		if err != nil {
+			return nil, Wrap(err, "error attempting challenge validation")
+		}
+		if ch.getStatus() == StatusValid {
+			break
+		}
+		if ch.getStatus() == StatusInvalid {
+			return ch.toACME(a.db, a.dir, p)
+		}
+		time.Sleep(ch.getBackoff())
```

Comment:
I would like to change this to a timer with context so we can cancel pending retries when a new request comes in. Also, a replicated CA needs to be able to communicate not just retry state with other CAs, but share the responsibility of executing retries. I _think_ I have a solution but I need to review the spec more to make sure there's a meaningful difference between a client re-post-ing to the challenge endoing and a client requiesting an entirely new validation.

Reply:
s/validation/_order_

Reply:
Upon review, there is definitely a difference between a re-requesting a challenge and requesting a new order.

My working strategy will be to have the server check the challenge record prior to executing a retry in order to make sure the current time is _after_ a `RetryAfter` field and the current Count is less than some maximum. If the RetryAfter field is in the past, prior to executing the retry, the server will update the RetryAfter field to be in the future and increment the Count field. 

If a server receives a new request on the challenge endpoint, it will first reset the count field on the challenge record, move RetryAfter into the future, and then immediately perform a validation attempt. Upon transient error, the server will schedule a retry at the time it listed in the RetryAfter field.

I think it also makes sense to add an identifier field so that the instance receiving the latest client challenge request can take over retries by updating the field to its own id.  When instances start, they query for all outstanding challenges with retries < MAX and ordinal == theirOrdinal. This enables graceful restart in a replicated context without one host sucking up all the retry responsibility. In k8s, you would deploy step-ca as a stateful set and pass the ordinal as this identifier.